### DISCUSSION TITLE
Can we manage clearing context using dynamic delayed transitions?

### DISCUSSION THREAD

**Discussion by hanzlahabib** [statelyai/xstate#2025] [Q&A]
I am trying to mimic the behaviour of user session timeout
here is scenario:
step 1:
user reads information and click next
step 2:
user submits details in form which are saved in xstate context and then press next
step 3:
now user is reading some terms and policy stuff, if user don't act with in specified mins, i want to clear user details which he submitted in step 2 and send this user back to the step 2,  

can i achieve this using dynamic delayed transitions?

Thanks

**Comment by Andarist** [comment]
You can achieve this using delayed transitions - not exactly sure what do u mean by dynamic delayed transitions though, so hard to answer that.

**Comment by hanzlahabib** [comment]
Dynamic transition means, I don't want to hard code the number of delayed
seconds, I receive timeout seconds from BE API

**Reply by Andarist** [reply]
Well, then yes - you could model this using dynamic delayed transitions.

**Comment by hanzlahabib** [comment]
How do you suggest to model delayed transition using dynamic variable?

**Reply by Andarist** [reply]
```js
after: [
  {
    delay: (ctx, ev) => { /* write your logic here */ },
    target: 'foo'
  }
]
```

**Reply by hanzlahabib** [reply]
Thanks @Andarist, i have another question
don't know if it is appropriate to ask here or create another question

i am persisting my xstate into LocalStorage and then restoring its state on initial load

I persist xstate only when `beforeunload` event is triggered, its brower unmounting its dom, 
before persisting xstate i send few actions to machine for cleanup and then persist clean machine into localStorage so that no API error is persisted which are in managed through context

what happens when i just send action to machine and persist state to localstorage, it don't wait until last action is completed,

for that i notice action sender returns state object which is updated copy of current state
to overcome this i stored the returned value of action emitter/creator to a new variable and then stored that to LocalStorage

it works fine for now, do you suggest any better way of doing this thing?
as i read it in xstate codebase, there was comment saying something like this `send action will return void after v5`,
